pkg/middleware/validate_action_url.go

```diff
@@ -0,0 +1,83 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"path"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+func ValidateActionUrl(cfg *setting.Cfg) web.Handler {
+	return func(c *contextmodel.ReqContext) {
+		// ignore local render calls
+		if c.IsRenderCall {
+			return
+		}
+
+		// only process requests targeting local instance
+		if !isLocalPath(c) {
+			// call next on url?
+			return
+		}
+
+		// only process POST and PUT
+		if c.Req.Method != http.MethodPost && c.Req.Method != http.MethodPut {
+			return
+		}
+
+		if c.IsApiRequest() {
+			// check if action header exists
+			action := c.Req.Header.Get("X-Grafana-Action")
+
+			if action == "" {
+				// header not found, just return
+				return
+			}
+
+			urlToCheck := c.Req.URL
```

Comment:
I think this wants to be `urlToCheck := path.Clean(c.Req.URL)` to prevent traversal attacks-- I don't see anything in the grafana request handling logic up to this point that would clean the path otherwise, although I've possibly missed something.

Reply:
sounds good, great catch!